bpo-30296 Remove unnecessary tuples, lists, sets, and dicts

jdufresne opened the pull request:
* Replaced list(<generator expression>) with list comprehension
* Replaced dict(<generator expression>) with dict comprehension
* Replaced set(<generator expression>) with set comprehension
* Replaced builtin func(<list comprehension>) with func(<generator
  expression>) when supported (e.g. any(), all(), tuple(), min(), &
  max())

jdufresne committed:
bpo-30296 Remove unnecessary tuples, lists, sets, and dicts

* Replaced list(<generator expression>) with list comprehension
* Replaced dict(<generator expression>) with dict comprehension
* Replaced set(<list literal>) with set literal
* Replaced builtin func(<list comprehension>) with func(<generator
  expression>) when supported (e.g. any(), all(), tuple(), min(), &
  max())

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://cpython-devguide.readthedocs.io/pullrequest.html#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

rhettinger commented on Lib/_weakrefset.py:
```diff
@@ -157,19 +157,19 @@ def issubset(self, other):
 
     def __lt__(self, other):
-        return self.data < set(ref(item) for item in other)
+        return self.data < {ref(item) for item in other}
```
Let's use:  set(map(ref, other)) for all of these.

rhettinger commented on Lib/distutils/msvc9compiler.py:
```diff
@@ -255,7 +255,7 @@ def query_vcvarsall(version, arch="x86"):
     """
     vcvarsall = find_vcvarsall(version)
-    interesting = set(("include", "lib", "libpath", "path"))
+    interesting = {"include", "lib", "libpath", "path"}
```
Yes to this one.  Set literals are always a win.

rhettinger commented on Lib/inspect.py:
```diff
@@ -368,7 +368,7 @@ def classify_class_attrs(cls):
     mro = getmro(cls)
     metamro = getmro(type(cls)) # for attributes stored in the metaclass
-    metamro = tuple([cls for cls in metamro if cls not in (type, object)])
+    metamro = tuple(cls for cls in metamro if cls not in (type, object))
```
Yes, this one is fine.

serhiy-storchaka replied:
I would keep the original code.

rhettinger commented on Lib/ntpath.py:
```diff
@@ -611,14 +611,14 @@ def commonpath(paths):
 
         try:
-            isabs, = set(p[:1] == sep for d, p in drivesplits)
+            isabs, = {p[:1] == sep for d, p in drivesplits}
```
This code smells bad (both the original and revision).  Can this be replaced with any(...).

rhettinger commented on Lib/optparse.py:
```diff
@@ -770,7 +770,7 @@ def convert_value(self, opt, value):
                 return self.check_value(opt, value)
             else:
-                return tuple([self.check_value(opt, v) for v in value])
+                return tuple(self.check_value(opt, v) for v in value)
```
This looks like a reasonable change, but optparse is essentially defunct so we should leave it alone.

rhettinger commented on Lib/pstats.py:
```diff
@@ -500,8 +500,8 @@ def add_callers(target, source):
                 # format used by cProfile
-                new_callers[func] = tuple([i[0] + i[1] for i in
-                                           zip(caller, new_callers[func])])
+                new_callers[func] = tuple(i[0] + i[1] for i in
```
Yes, this is okay.

rhettinger commented on Lib/pstats.py:
```diff
@@ -500,8 +500,8 @@ def add_callers(target, source):
-                new_callers[func] = tuple([i[0] + i[1] for i in
-                                           zip(caller, new_callers[func])])
+                new_callers[func] = tuple(i[0] + i[1] for i in
+                                          zip(caller, new_callers[func]))
```
If the line isn't too long, this is okay.

rhettinger commented on Lib/sqlite3/dbapi2.py:
```diff
@@ -47,8 +47,8 @@ def TimeFromTicks(ticks):
-version_info = tuple([int(x) for x in version.split(".")])
-sqlite_version_info = tuple([int(x) for x in sqlite_version.split(".")])
+version_info = tuple(int(x) for x in version.split("."))
+sqlite_version_info = tuple(int(x) for x in sqlite_version.split("."))
```
The suggest change is a minor improvement, but this code is very old and is probably best left untouched.

rhettinger commented on Lib/symtable.py:
```diff
@@ -119,8 +119,8 @@ class Function(SymbolTable):
     def __idents_matching(self, test_func):
-        return tuple([ident for ident in self.get_identifiers()
-                      if test_func(self._table.symbols[ident])])
+        return tuple(ident for ident in self.get_identifiers()
```
This is okay.

serhiy-storchaka replied:
I would keep the original code.

rhettinger commented on Lib/tkinter/filedialog.py:
```diff
@@ -330,7 +330,7 @@ class Open(_Dialog):
         if isinstance(result, tuple):
             # multiple results:
-            result = tuple([getattr(r, "string", r) for r in result])
+            result = tuple(getattr(r, "string", r) for r in result)
```
This is a file I prefer not to touch.  Otherwise the change itself seems reasonable.

rhettinger commented on Lib/tokenize.py:
```diff
@@ -142,7 +142,7 @@ def _all_string_prefixes():
     _valid_string_prefixes = ['b', 'r', 'u', 'f', 'br', 'fr']
     # if we add binary f-strings, add: ['fb', 'fbr']
-    result = set([''])
+    result = {''}
```
This is okay.  Set literals are almost always the right thing to do.  My only misgiving is that the new code looks like an ascii emoticon.

rhettinger commented on Lib/traceback.py:
```diff
@@ -253,8 +253,7 @@ def __init__(self, filename, lineno, name, *, lookup_line=True,
             self.line
-        self.locals = \
-            dict((k, repr(v)) for k, v in locals.items()) if locals else None
+        self.locals = {k: repr(v) for k, v in locals.items()} if locals else None
```
This is okay.

rhettinger commented on Lib/turtle.py:
```diff
@@ -1175,7 +1175,7 @@ def _color(self, cstr):
         else:
             raise TurtleGraphicsError("bad colorstring: %s" % cstr)
-        return tuple([c * self._colormode/255 for c in cl])
+        return tuple(c * self._colormode/255 for c in cl)
```
This is okay.

serhiy-storchaka replied:
I would keep the original code.

rhettinger commented on Lib/turtle.py:
```diff
@@ -2989,7 +2989,7 @@ def _getshapepoly(self, polygon, compound=False):
         elif self._resizemode == "noresize":
             return polygon
-        return tuple([(t11*x + t12*y, t21*x + t22*y) for (x, y) in polygon])
+        return tuple((t11*x + t12*y, t21*x + t22*y) for (x, y) in polygon)
```
This is okay.

serhiy-storchaka replied:
I would keep the original code.

rhettinger commented on Lib/turtle.py:
```diff
@@ -3839,8 +3839,8 @@ def write_docstringdict(filename="turtle_docstringdict"):
     with open("%s.py" % filename,"w") as f:
-        keys = sorted([x for x in docsdict.keys()
-                            if x.split('.')[1] not in _alias_list])
+        keys = sorted(x for x in docsdict.keys()
```
This is okay.

serhiy-storchaka replied:
`.keys()` can be removed. Iterate just `docsdict`.

rhettinger commented on Lib/turtledemo/wikipedia.py:
```diff
@@ -52,7 +52,7 @@ def main():
 
     at = clock()
-    while any([t.undobufferentries() for t in s.turtles()]):
+    while any(t.undobufferentries() for t in s.turtles()):
```
This is okay.

rhettinger commented on Lib/urllib/request.py:
```diff
@@ -845,7 +845,7 @@ def add_password(self, realm, uri, user, passwd):
         for default_port in True, False:
             reduced_uri = tuple(
-                [self.reduce_uri(u, default_port) for u in uri])
+                self.reduce_uri(u, default_port) for u in uri)
```
This is okay

serhiy-storchaka replied:
I would keep the original code.

rhettinger commented on Lib/urllib/request.py:
```diff
@@ -1286,8 +1286,7 @@ def do_open(self, http_class, req, **http_conn_args):
         headers = dict(req.unredirected_hdrs)
-        headers.update(dict((k, v) for k, v in req.headers.items()
-                            if k not in headers))
+        headers.update({k: v for k, v in req.headers.items() if k not in headers})
```
update() can accept an iterable of tuples, so change this to:
headers.update((k, v) for k, v in req.headers.items() if k not in headers)

rhettinger commented on Lib/urllib/request.py:
```diff
@@ -1299,7 +1298,7 @@ def do_open(self, http_class, req, **http_conn_args):
         # request.
         headers["Connection"] = "close"
-        headers = dict((name.title(), val) for name, val in headers.items())
+        headers = {name.title(): val for name, val in headers.items()}
```
This is okay.  Sometimes the dict() form is clearer about its intention but in this case the colon form for the key/value is clearer than the (key, value) form.

rhettinger commented on Mac/BuildScript/build-installer.py:
```diff
@@ -84,7 +84,7 @@ def getVersion():
 
 def getVersionMajorMinor():
-    return tuple([int(n) for n in getVersion().split('.', 2)])
+    return tuple(int(n) for n in getVersion().split('.', 2))
```
This would be okay but it is in a file that we usually avoid touching.

rhettinger commented on Tools/gdb/libpython.py:
```diff
@@ -1097,8 +1097,8 @@ def proxyval(self, visited):
-        result = tuple([PyObjectPtr.from_pyobject_ptr(self[i]).proxyval(visited)
-                        for i in safe_range(int_from_int(self.field('ob_size')))])
+        result = tuple(PyObjectPtr.from_pyobject_ptr(self[i]).proxyval(visited)
+                       for i in safe_range(int_from_int(self.field('ob_size'))))
```
This is okay.

serhiy-storchaka replied:
I would keep the original code.

rhettinger commented on Tools/scripts/byext.py:
```diff
@@ -83,7 +83,7 @@ def report(self):
         cols = sorted(columns)
         colwidth = {}
-        colwidth["ext"] = max([len(ext) for ext in exts])
+        colwidth["ext"] = max(len(ext) for ext in exts)
```
This is okay but I like max(map(len, exts)) better

rhettinger commented on Tools/unicode/makeunicodedata.py:
```diff
@@ -609,7 +609,7 @@ def makeunicodename(unicode, trace):
                 names[char] = name + chr(0)
 
-    print(len(list(n for n in names if n is not None)), "distinct names")
+    print(len([n for n in names if n is not None]), "distinct names")
```
This is okay

rhettinger commented on Lib/_compat_pickle.py:
```diff
@@ -162,9 +162,9 @@
 
 # Same, but for 3.x to 2.x
-REVERSE_IMPORT_MAPPING = dict((v, k) for (k, v) in IMPORT_MAPPING.items())
+REVERSE_IMPORT_MAPPING = {v: k for k, v in IMPORT_MAPPING.items()}
```
Lets skip this module.

rhettinger commented on Lib/_pydecimal.py:
```diff
@@ -3952,14 +3952,14 @@ def __init__(self, prec=None, rounding=None, Emin=None, Emax=None,
             self.traps = dc.traps.copy()
         elif not isinstance(traps, dict):
-            self.traps = dict((s, int(s in traps)) for s in _signals + traps)
+            self.traps = {s: int(s in traps) for s in _signals + traps}
```
Let's skip this module.

rhettinger commented on Lib/argparse.py:
```diff
@@ -263,7 +263,7 @@ def add_argument(self, action):
 
             # update the maximum item length
-            invocation_length = max([len(s) for s in invocations])
+            invocation_length = max(len(s) for s in invocations)
```
Let's skip this module.  I think Bethard would like this to be compatible across as many versions as possible or at least to keep the code in-sync as much as possible.

rhettinger commented on Lib/ast.py:
```diff
@@ -56,14 +56,14 @@ def _convert(node):
             return node.n
         elif isinstance(node, Tuple):
-            return tuple(map(_convert, node.elts))
+            return tuple(_convert(v) for v in node.elts)
```
In all of these cases, I prefer the original for clarity and inter-line parallelism

rhettinger commented on Lib/cgi.py:
```diff
@@ -644,7 +644,7 @@ def keys(self):
         if self.list is None:
             raise TypeError("not indexable")
-        return list(set(item.name for item in self.list))
+        return list({item.name for item in self.list})
```
Let's skip this one.  I think the original is clearer about its intention.

rhettinger commented on Lib/cmd.py:
```diff
@@ -285,8 +285,8 @@ def get_names(self):
         commands = set(self.completenames(*args))
-        topics = set(a[5:] for a in self.get_names()
-                     if a.startswith('help_' + args[0]))
+        topics = {a[5:] for a in self.get_names()
```
Skip this one.  The original is clearer about its intentions.

rhettinger commented on Lib/concurrent/futures/_base.py:
```diff
@@ -193,9 +193,9 @@ def as_completed(fs, timeout=None):
     fs = set(fs)
     with _AcquireFutures(fs):
-        finished = set(
+        finished = {
```
Let's skip all of these.  This module is delicate (harder to test minor changes) and the original code is marginally a bit clearer.

rhettinger commented on Lib/distutils/cygwinccompiler.py:
```diff
@@ -397,7 +397,7 @@ def get_versions():
     """
     commands = ['gcc -dumpversion', 'ld -v', 'dllwrap --version']
-    return tuple([_find_exe_version(cmd) for cmd in commands])
+    return tuple(_find_exe_version(cmd) for cmd in commands)
```
This change would be okay but it is in a module I prefer not to touch.

rhettinger commented on Lib/email/headerregistry.py:
```diff
@@ -369,8 +369,8 @@ def groups(self):
+            self._addresses = tuple(address for group in self._groups
+                                            for address in group.addresses)
         return self._addresses
 
```
This is okay.

serhiy-storchaka replied:
The original variant, with list comprehension, is faster than the variant with a generator. The difference is up to 2 times for simple tests.

I would keep the original code.

rhettinger commented on Lib/idlelib/multicall.py:
```diff
@@ -134,7 +134,7 @@ def nbits(n):
     statelist = []
     for state in states:
-        substates = list(set(state & x for x in states))
+        substates = list({state & x for x in states})
```
Skip this one.  I think the original is clearer.

terryjreedy replied:
Agree with Raymond, leave as is.

rhettinger commented on Lib/idlelib/multicall.py:
```diff
@@ -52,9 +52,9 @@
-                         for name in _modifiers[number]])
+_modifier_names = {name: number
+                   for number in range(len(_modifiers))
+                   for name in _modifiers[number]}
```
This is okay.

terryjreedy replied:
I prefer the change as easier to read and likely faster.  Ditto for the other dict(list) to dict comp change in this file.  (Agreeing with Raymond.)

rhettinger commented on Lib/idlelib/multicall.py:
```diff
@@ -258,9 +258,9 @@ def __del__(self):
-                     for name in _types[number]])
+_type_names = {name: number
+               for number in range(len(_types))
+               for name in _types[number]}
```
This is okay.

rhettinger commented on Lib/lib2to3/main.py:
```diff
@@ -218,7 +218,7 @@ def main(fixer_pkg, args=None):
     # Initialize the refactoring tool
     avail_fixes = set(refactor.get_fixers_from_package(fixer_pkg))
-    unwanted_fixes = set(fixer_pkg + ".fix_" + fix for fix in options.nofix)
+    unwanted_fixes = {fixer_pkg + ".fix_" + fix for fix in options.nofix}
```
Skip this.  I prefer the original for clarity and for parallelism with the previous line.

rhettinger commented on Lib/logging/config.py:
```diff
@@ -463,7 +463,7 @@ def configure_custom(self, config):
         props = config.pop('.', None)
         # Check for valid identifiers
-        kwargs = dict([(k, config[k]) for k in config if valid_ident(k)])
+        kwargs = {k: config[k] for k in config if valid_ident(k)}
```
This should use the iterable-of-pairs form instead of creating an intermediate dictionary:
dict((k, config[k]) for k in config if valid_ident(k))

rhettinger commented on Lib/logging/config.py:
```diff
@@ -726,7 +726,7 @@ def configure_handler(self, config):
             factory = klass
         props = config.pop('.', None)
-        kwargs = dict([(k, config[k]) for k in config if valid_ident(k)])
+        kwargs = {k: config[k] for k in config if valid_ident(k)}
```
Same comment here:
dict((k, config[k]) for k in config if valid_ident(k)).
The proposed change also reads well but I'm not sure whether Vinay would want dict comps in this module.

rhettinger commented on Lib/multiprocessing/context.py:
```diff
@@ -261,7 +261,7 @@ def get_all_start_methods(self):
                 return ['fork', 'spawn']
 
-DefaultContext.__all__ = list(x for x in dir(DefaultContext) if x[0] != '_')
+DefaultContext.__all__ = [x for x in dir(DefaultContext) if x[0] != '_']
```
This is okay.

rhettinger commented on Lib/multiprocessing/forkserver.py:
```diff
@@ -98,8 +98,7 @@ def ensure_running(self):
                 data = spawn.get_preparation_data('ignore')
-                data = dict((x,y) for (x,y) in data.items()
-                            if x in desired_keys)
+                data = {x: y for x, y in data.items() if x in desired_keys}
```
I'm ambivalent about this one.

rhettinger commented on Lib/multiprocessing/sharedctypes.py:
```diff
@@ -115,7 +115,7 @@ def synchronized(obj, lock=None, ctx=None):
         except KeyError:
             names = [field[0] for field in cls._fields_]
-            d = dict((name, make_property(name)) for name in names)
+            d = {name: make_property(name) for name in names}
```
This is okay.

rhettinger commented on Lib/ntpath.py:
```diff
@@ -611,14 +611,14 @@ def commonpath(paths):
         # now otherwise type errors for mixing strings and bytes would not be
         # caught.
-        if len(set(d for d, p in drivesplits)) != 1:
+        if len({d for d, p in drivesplits}) != 1:
```
I prefer the original for clarity.

rhettinger commented on Lib/pathlib.py:
```diff
@@ -115,8 +115,8 @@ class _WindowsFlavour(_Flavour):
     drive_letters = (
-        set(chr(x) for x in range(ord('a'), ord('z') + 1)) |
-        set(chr(x) for x in range(ord('A'), ord('Z') + 1))
+        {chr(x) for x in range(ord('a'), ord('z') + 1)} |
```
This should be simplified to set('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ')
or if string is already imported to set(string.ascii_letters)

rhettinger commented on Lib/posixpath.py:
```diff
@@ -500,7 +500,7 @@ def commonpath(paths):
 
         try:
-            isabs, = set(p[:1] == sep for p in paths)
+            isabs, = {p[:1] == sep for p in paths}
```
This code smells bad.  Is there a better way to express what it is trying to do?

jdufresne replied:
Perhaps. However, my intention with this change was to avoid logic changes. I'll revert `posixpath.py` and `ntpath.py` for now.

rhettinger commented on Lib/signal.py:
```diff
@@ -58,15 +58,15 @@ def getsignal(signalnum):
     def pthread_sigmask(how, mask):
         sigs_set = _signal.pthread_sigmask(how, mask)
-        return set(_int_to_enum(x, Signals) for x in sigs_set)
+        return {_int_to_enum(x, Signals) for x in sigs_set}
```
Skip this, I prefer the original.

rhettinger commented on Lib/signal.py:
```diff
@@ -58,15 +58,15 @@ def getsignal(signalnum):
     def sigpending():
         sigs = _signal.sigpending()
-        return set(_int_to_enum(x, Signals) for x in sigs)
+        return {_int_to_enum(x, Signals) for x in sigs}
```
Skip this. I prefer the original.

rhettinger commented on Lib/urllib/request.py:
```diff
@@ -683,8 +683,8 @@ def redirect_request(self, req, fp, code, msg, headers, newurl):
-        newheaders = dict((k, v) for k, v in req.headers.items()
-                          if k.lower() not in CONTENT_HEADERS)
+        newheaders = {k: v for k, v in req.headers.items()
+                      if k.lower() not in CONTENT_HEADERS}
```
I'm ambivalent about this one.

rhettinger commented on Tools/msi/csv_to_wxs.py:
```diff
@@ -44,7 +44,7 @@ def main(file_source, install_target):
         files = list(csv.reader(f))
 
-    assert len(files) == len(set(make_id(f[1]) for f in files)), "Duplicate file IDs exist"
+    assert len(files) == len({make_id(f[1]) for f in files}), "Duplicate file IDs exist"
```
Skip this.  The original is clearer and the fix is one we usually don't touch unless we have to.

rhettinger commented on Tools/ssl/make_ssl_data.py:
```diff
@@ -47,8 +47,8 @@ def parse_error_codes(h_file, prefix, libcode):
                                        prefix, libcode)
-    new_code_nums = set((libcode, num)
-                        for (code, (libcode, name, num)) in new_codes)
+    new_code_nums = {(libcode, num)
```
Skip this.  I prefer the original and the file is one I prefer not to touch.

rhettinger commented on Tools/stringbench/stringbench.py:
```diff
@@ -1323,8 +1323,8 @@ def quick_replace_multiple_match(STR):
 _format_unicode = unicode_from_str(_format)
-_format_dict_bytes = dict((bytes_from_str(k), bytes_from_str(v)) for (k,v) in _format_dict.items())
-_format_dict_unicode = dict((unicode_from_str(k), unicode_from_str(v)) for (k,v) in _format_dict.items())
+_format_dict_bytes = {bytes_from_str(k): bytes_from_str(v) for k, v in _format_dict.items()}
```
I prefer not to touch a benchmark file lest it cause comparability issues.

rhettinger commented on setup.py:
```diff
@@ -206,7 +206,7 @@ def build_extensions(self):
                       if ext.name not in disabled_module_list]
         # move ctypes to the end, it depends on other modules
-        ext_map = dict((ext.name, i) for i, ext in enumerate(extensions))
+        ext_map = {ext.name: i for i, ext in enumerate(extensions)}
```
Skip this.  I prefer not to touch anything in setup.py.

rhettinger commented on setup.py:
```diff
@@ -271,10 +271,10 @@ def build_extensions(self):
             self.check_extension_import(ext)
 
-        longest = max([len(e.name) for e in self.extensions], default=0)
+        longest = max((len(e.name) for e in self.extensions), default=0)
```
Skip

rhettinger commented on setup.py:
```diff
@@ -271,10 +271,10 @@ def build_extensions(self):
         if self.failed or self.failed_on_import:
             all_failed = self.failed + self.failed_on_import
-            longest = max(longest, max([len(name) for name in all_failed]))
+            longest = max(longest, max(len(name) for name in all_failed))
```
Skip

rhettinger commented on setup.py:
```diff
@@ -1162,8 +1162,7 @@ class db_found(Exception): pass
                     sqlite_version = m.group(1)
-                    sqlite_version_tuple = tuple([int(x)
-                                        for x in sqlite_version.split(".")])
+                    sqlite_version_tuple = tuple(int(x) for x in sqlite_version.split("."))
```
Ambivalent.

terryjreedy reviewed (changes requested):
I keep idlelib 3.6 and 3.7 in sync as much as possible.  
Please remove all Lib/idlelib/multicall.py changes from this PR and optionally (and preferably) make the 2 approved changes in a separate PR that can be backported easily.

jdufresne commented:
Thanks for the reviews! I believe I have addressed all feedback comments. If I missed or misunderstood any of the previous comments, please let me know.

serhiy-storchaka commented on Lib/logging/config.py:
```diff
@@ -463,7 +463,7 @@ def configure_custom(self, config):
         props = config.pop('.', None)
         # Check for valid identifiers
-        kwargs = dict([(k, config[k]) for k in config if valid_ident(k)])
+        kwargs = dict((k, config[k]) for k in config if valid_ident(k))
```
I would use a dict comprehension.

And the same for other change in this file.

serhiy-storchaka commented on Lib/pstats.py:
```diff
@@ -500,8 +500,7 @@ def add_callers(target, source):
                 # format used by cProfile
-                new_callers[func] = tuple([i[0] + i[1] for i in
-                                           zip(caller, new_callers[func])])
+                new_callers[func] = tuple(i[0] + i[1] for i in zip(caller, new_callers[func]))
```
I would keep the original code.

Maybe unpack a tuple:
```
tuple([i + j for i, j in zip(caller, new_callers[func])])
```

serhiy-storchaka commented on Lib/urllib/request.py:
```diff
@@ -1286,8 +1286,7 @@ def do_open(self, http_class, req, **http_conn_args):
         headers = dict(req.unredirected_hdrs)
-        headers.update(dict((k, v) for k, v in req.headers.items()
-                            if k not in headers))
+        headers.update((k, v) for k, v in req.headers.items() if k not in headers)
```
This changes the semantic. Updated `headers` is used in the generator expression. I don't know whether this is good.

I would use a dict comprehesion.
